precompiles/gmp/src/lib.rs

```diff
@@ -0,0 +1,312 @@
+				.build(),
+		)?;
+		let mut reader = EvmDataReader::new(&output[..]);
+		let before_amount: U256 = reader.read()?;
+		log::debug!(target: "gmp-precompile", "before balance: {}", before_amount);
+
+		// our inner-most payload should be a VersionedUserAction
+		let user_action = VersionedUserAction::decode_with_depth_limit(
+			32,
+			&mut transfer_with_payload.payload.as_bytes(),
+		)
+		.map_err(|_| RevertReason::Custom("Invalid GMP Payload".into()))?;
+		log::debug!(target: "gmp-precompile", "user action: {:?}", user_action);
+
+		let currency_account_id = Runtime::AddressMapping::into_account_id(wrapped_address.into());
+
+		let currency_id: <Runtime as orml_xtokens::Config>::CurrencyId =
+			Runtime::account_to_currency_id(currency_account_id)
+				.ok_or(revert("Unsupported asset, not a valid currency id"))?;
+
+		let amount = transfer_with_payload
+			.amount
+			.try_into()
+			.map_err(|_| revert("Amount overflows balance"))?;
+
+		log::debug!(target: "gmp-precompile", "attempt to transfer {:?} of {:?}", amount, currency_id);
+
+		// Complete a "Contract Controlled Transfer" with the given Wormhole VAA.
+		// We need to invoke Wormhole's completeTransferWithPayload function, passing it the VAA.
+		// Upon success, it should have transferred tokens to this precompile's address.
+		Self::call(
+			handle,
+			wormhole_bridge,
+			EvmDataWriter::new_with_selector(COMPLETE_TRANSFER_WITH_PAYLOAD_SELECTOR)
+				.write(wormhole_vaa)
+				.build(),
+		)?;
+
+		// query our "after" balance (our being this precompile)
+		let output = Self::call(
+			handle,
+			wrapped_address.into(),
+			EvmDataWriter::new_with_selector(BALANCE_OF_SELECTOR)
+				.write(Address::from(handle.code_address()))
+				.build(),
+		)?;
+		let mut reader = EvmDataReader::new(&output[..]);
+		let after_amount: U256 = reader.read()?;
+		log::debug!(target: "gmp-precompile", "after balance: {}", after_amount);
+
+		let amount_transfered = after_amount.saturating_sub(before_amount);
+
+		// TODO: review
+		if amount_transfered < transfer_with_payload.amount {
+			log::debug!(
+				target: "gmp-precompile",
+				"insufficient funds bridged ({} < {})",
+				amount_transfered, transfer_with_payload.amount);
+			return Err(PrecompileFailure::Revert {
+				exit_status: ExitRevert::Reverted,
+				output: "insufficient funds bridged".into(),
+			});
+		}
+
+		// TODO: Wormhole might have transfered unsupported tokens; we should handle this case
+		//       gracefully (maybe that's as simple as reverting)
+
+		// TODO:
+		let weight_limit: u64 = 1_000_000_000_000u64;
+
+		log::debug!(target: "gmp-precompile", "sending XCM via xtokens::transfer...");
+		let call: orml_xtokens::Call<Runtime> = match user_action {
+			VersionedUserAction::V1(action) => orml_xtokens::Call::<Runtime>::transfer {
+				currency_id,
+				amount,
+				dest: Box::new(VersionedMultiLocation::V3(action.destination)),
+				dest_weight_limit: WeightLimit::Limited(Weight::from_parts(
```

Comment:
Use the Unlimited variant instead

Reply:
The `Limited` variant should be used only if the user/dapp has specified a limit